frontend/packages/console-app/src/components/detect-namespace/namespace.ts

```diff
@@ -3,90 +3,89 @@ import * as React from 'react';
-  );
-  const [lastNamespace, setLastNamespace, lastNamespaceLoaded] = useUserSettingsCompatibility<
-    string
-  >(LAST_NAMESPACE_NAME_USER_SETTINGS_KEY, LAST_NAMESPACE_NAME_LOCAL_STORAGE_KEY);
-
+  const [lastNamespace, setLastNamespace, lastNamespaceLoaded] = useLastNamespace();
+  const [preferredNamespace, preferredNamespaceLoaded] = usePreferredNamespace();
+  const [activeNamespace, setActiveNamespace] = React.useState<string>('');
   const dispatch = useDispatch();
   const setNamespace = React.useCallback(
     (namespace: string) => {
-      dispatch(setActiveNamespace(namespace));
+      setActiveNamespace(namespace);
+      dispatch(setActiveNamespaceForStore(namespace));
       setLastNamespace(namespace);
     },
     [dispatch, setLastNamespace],
   );
-  const useProjects = useFlag(FLAGS.OPENSHIFT);
+  const useProjects: boolean = useFlag(FLAGS.OPENSHIFT);
+
+  const resourcesLoaded: boolean = !!(
+    urlNamespace ||
+    (preferredNamespaceLoaded && lastNamespaceLoaded)
+  );
 
-  // Keep namespace in sync with redux.
   React.useEffect(() => {
-    // Url namespace overrides everything. We don't need to wait for loaded
-    if (urlNamespace) {
-      dispatch(setActiveNamespace(urlNamespace));
-    }
-    // Automatically sets favorited or latest namespace as soon as
-    // both informations are loaded from user settings.
-    if (!urlNamespace && favoriteLoaded && lastNamespaceLoaded && !flagPending(useProjects)) {
-      const setActiveIfExists = (ns: string) => {
-        if (!ns) {
-          return Promise.reject();
-        }
-        return namespaceExists(useProjects ? ProjectModel : NamespaceModel, ns)
-          .then(() => dispatch(setActiveNamespace(ns)))
-          .catch((res) => {
-            // eslint-disable-next-line no-console
-            console.warn('Error fetching namespace', ns);
-            throw res;
-          });
-      };
-      setActiveIfExists(favoritedNamespace)
-        .catch(() => setActiveIfExists(lastNamespace))
-        .catch(() => {});
+    if (resourcesLoaded) {
+      getValueForNamespace(
+        useProjects,
+        urlNamespace,
+        activeNamespace,
+        preferredNamespace,
+        lastNamespace,
+      )
+        .then((ns) => {
+          if (ns !== activeNamespace) {
+            setActiveNamespace(ns);
+            // sync with redux store
+            dispatch(setActiveNamespaceForStore(ns));
+          }
+        })
+        .catch((e) => {
+          // eslint-disable-next-line no-console
+          console.warn('Error fetching namespace', e);
+        });
     }
-    // Only run this hook after favorite and last namespace are loaded.
+    // Only run this hook after all resources have loaded.
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [favoriteLoaded, lastNamespaceLoaded, useProjects]);
+  }, [resourcesLoaded, useProjects]);
+
+  React.useEffect(() => {
+    if (urlNamespace) {
+      getValueForNamespace(useProjects, urlNamespace)
```

Comment:
Don't you also need to wait for `!flagPending(useProjects)` ?

Reply:
yes. added it to `resourceLoaded`.
Thanks for catching this.